Report conflicting keys

zjuLcg opened the pull request:
This feature has already been implemented and tested.

Current implementation:
Using special keys through `getRange(\xff\xff/transaction/conflicting_keys/, \xff\xff/transaction/conflicting_keys/\xff)` to retrieve all conflicting keys. The returned key/value pairs are formatted as:

Key: any valid keys with the prefix
Value: a boolean('0' or '1') that represents whether keys greater or equal than the key is conflicting.

E.g.
```
'\xff\xff/transaction/conflicting_keys/far': '1'
'\xff\xff/transaction/conflicting_keys/food': '0'
'\xff\xff/transaction/conflicting_keys/fool': '1'
`\xff\xff/transaction/conflicting_keys/fool\x00`: '0'
```
means two keyRange `[far, food)` and `[fool, fool\x00)` are causing conflicts. In each of them, at least one key is read by the transaction itself but also written by others(with higher `version`) at the same time. In particular, only one key `fool` in `[fool, fool\x00)`, which means `fool` is read and written at the same time and caused conflicts. 


However,  to add as small overhead to the resolver as possible, currently, the resolvers only return indices of `read_conflict_range`s that are conflicting. Thus, the returned `keyRange`s are the same as the users added to the transaction when they committed the transaction. 

A simple python [script](https://github.com/apple/foundationdb/pull/2257#issuecomment-561985787) to show the usage.

zjuLcg commented:
> 1. I think reporting a single key (keyrange?) that caused a conflict is probably sufficient for an application developer to debug their workload/data model. I'm not sure additional keys add much value, but I might be wrong here. @alecgrieser what do you think?

I agree with that one keyrange is enough for it. Made it larg

zjuLcg commented:
The basic code path for this follows:
``` 
commit -> commitAndWatch -> commitMutations -> tryCommit -> RequestStream<CommitTransactionRequest> -> commitBatcher -> PromiseStream<pair<vector<CommitTransactionRequest>, int>> -> 
commitBatch ->
RequestStream<ResolveTransactionBatchRequest> ->
resolveBatch
```

zjuLcg commented on fdbserver/MasterProxyServer.actor.cpp:
```diff
@@ -998,7 +1000,18 @@ ACTOR Future<Void> commitBatch(
+				for (int resolverInd = 0; resolverInd < resolution.size(); ++resolverInd){
+					for (auto & kr : resolution[resolverInd].conflictingKeyRangeMap[t]){
+						// TODO : Each resolver may give some same conflicting keys, do we need to cluster them?
+						TraceEvent("ConflictingKeyRange").detail("ResolverIndex", resolverInd).detail("TrasactionIndex", t).detail("StartKey", kr.begin.toString()).detail("EndKey", kr.end.toString());
```
Just use it for debugging right now. Will remove once done.

zjuLcg commented on fdbserver/ConflictSet.h:
```diff
@@ -33,7 +33,7 @@ void clearConflictSet( ConflictSet*, Version );
 
 struct ConflictBatch {
-	explicit ConflictBatch( ConflictSet* );
+	explicit ConflictBatch( ConflictSet*,  std::map<int, std::vector< KeyRange >>* conflictingKeyRangeMap = NULL);
```
Done. Thanks.

zjuLcg commented:
> Resolver is a place where we're likely to be pretty uncomfortable in regressing on CPU usage, because as soon as we hit 100% CPU usage on a resolver, we have to run multiple of them and have imprecise conflict checking as a result. I'm a little bit concerned about the allocations from std::map<std::vector<>> causing a noticable slowdown on transactions that have this enabled.
> 
> It'd definitely be a good idea to make a resolver stress test in circus (many keys, large keys, small/no values) to be able to measure the impact of your change on performance.

Thanks, Alex. Working on both correctness and performance testing.

alecgrieser commented:
From @atn34: 

> I think reporting a single key (keyrange?) that caused a conflict is probably sufficient for an application developer to debug their workload/data model. I'm not sure additional keys add much value, but I might be wrong here. @alecgrieser what do you think?

Hm, I suppose if it were free, adding additional key ranges would be nice to have, but yeah, having a single key range would probably be sufficient. The main use case, I think, should be that one can make sense of what changed.

I think key range over key. I'm trying to think whether it should return the modified range or the read range (if that makes sense). Perhaps both? (Like, reading range "a to c" conflicts with a change to range "b to d".) If only some of this is possible because of performance or what data are in the resolver, then some subset is probably sufficient.

alexmiller-apple commented:
> reading range "a to c" conflicts with a change to range "b to d"

In case we end up implementing finer grained ACLs at some point, we might wish to enforce that this API shouldn't return a key outside of the range that it read or wrote (and was permitted to read and write)?

ajbeamon commented:
> > reading range "a to c" conflicts with a change to range "b to d"
> 
> In case we end up implementing finer grained ACLs at some point, we might wish to enforce that this API shouldn't return a key outside of the range that it read or wrote (and was permitted to read and write)?

There's probably value to including the full details of the other range, and without an ACL implementation available it may be hard to do anything that's not too restrictive (i.e. that checks against the more lenient permitted ranges vs. the likely stricter modified ranges).

ajbeamon commented on fdbclient/NativeAPI.actor.cpp:
```diff
@@ -2784,6 +2788,10 @@ Future<Void> Transaction::commitMutations() {
 		}
+		if(options.reportConflictingKeys) {
+			// TODO : Is it better to keep it as a flag?
+			tr.reportConflictingKeys();
```
I think we have other cases where we access the flag directly out of options rather than copy it to the transaction, so that would probably be my preference.

zjuLcg replied:
Not sure if I understand correctly.
I keep it in `CommitTransactionRef` since `CommitTransactionRef` will be directly copied to `ResolveTransactionBatchRequest` later.
If I use `flag` here, it is in `CommitTransactionRequest`, which needs a further step to be copied to `ResolveTransactionBatchRequest`.

zjuLcg commented:
@alexmiller-apple @ajbeamon @alecgrieser 
Do you think it is okay to make the report-conflicting-key API as a simple call like `tr.get("\xff\xff/conflicting_keys/json")`, which can avoid compatibility issues? 
And the return value can be a `json` string like what `\xff\xff/status/json` returns. 
Since a general use case may be a user calls the API after several times of retries but always gets conflicts, it can be formatted as each `commit` corresponding to its conflicting keys and other information if needed.
Not sure if I miss anything. Any thoughts or advice will be a great help for me. Thanks.

zjuLcg commented on fdbclient/ReadYourWrites.actor.cpp:
```diff
@@ -1229,17 +1229,23 @@ Future< Optional<Value> > ReadYourWritesTransaction::get( const Key& key, bool s
-			for (auto & cKR : tr.info.conflictingKeyRanges) {
-				for (auto & kr : cKR) {
-					conflictingKeysArray.push_back(format("[%s, %s)", kr.begin.toString().c_str(), kr.end.toString().c_str()));
+			json_spirit::mArray root;
```
Changed to use json_spirit to read and write json.

atn34 commented on fdbserver/workloads/ReportConflictingKeys.actor.cpp:
```diff
@@ -0,0 +1,166 @@
+#include "fdbclient/NativeAPI.actor.h"
```
This file probably needs the Apache 2 license header thing at the top, and please apply clang-format to it.

zjuLcg replied:
Will do. Thanks.

atn34 commented on tests/CMakeLists.txt:
```diff
@@ -65,6 +65,7 @@ add_fdb_test(TEST_FILES ReadHalfAbsent.txt IGNORE)
 add_fdb_test(TEST_FILES RedwoodCorrectnessUnits.txt IGNORE)
 add_fdb_test(TEST_FILES RedwoodCorrectnessBTree.txt IGNORE)
 add_fdb_test(TEST_FILES RedwoodCorrectnessPager.txt IGNORE)
+add_fdb_test(TEST_FILES ReportConflictingKeys.txt IGNORE)
```
This test shouldn't be ignored

zjuLcg replied:
Sure. Will change to basic test.

atn34 commented on fdbclient/CommitTransaction.h:
```diff
@@ -161,6 +163,7 @@ struct CommitTransactionRef {
 	}
 
 	void clear( Arena& arena, KeyRangeRef const& keys ) {
+		// TODO: check do I need to clear flag here
```
You don't need to clear the report_conflicting_keys flag here

zjuLcg replied:
Thanks. Forgot to delete this comment.

atn34 commented on fdbclient/NativeAPI.actor.cpp:
```diff
@@ -2784,6 +2788,11 @@ Future<Void> Transaction::commitMutations() {
+		if(options.reportConflictingKeys) {
+			// TODO : Is it better to keep it as a flag?
+			tr.flags = tr.flags | CommitTransactionRequest::FLAG_REPORT_CONFLICTING_KEYS;
+			tr.reportConflictingKeys();
```
There seems to be two (equivalent?) ways of indicating that we want to report conflict keys in the commit request. Let's just pick one.

zjuLcg replied:
Thanks. Cleaned it.

atn34 commented on fdbserver/workloads/ReportConflictingKeys.actor.cpp:
```diff
@@ -0,0 +1,166 @@
+                    isConflict = true;
+                }
+                wait(tr.onError(e));
+                // check API correctness
```
This looks good - I still want to get buy-in for what the exact semantics/contract of this feature should be from the layers folks before shipping though

atn34 replied:
I'm now thinking that we want to report a key range which intersects both the read conflict range of this transaction and the write conflict range of some other transaction, if recovering the exact write conflict range of the other transaction isn't feasible. Does that sound reasonable?

zjuLcg replied:
This makes sense. Has changed to narrow the returned `KeyRange` to be the intersection of original `readConflictRange` and some `writeConflictRange`.

zjuLcg commented:
As @alexmiller-apple and @atn34 suggested, I added a simple workload, `ReportConflictingKeysWorkload`, used to test both the `correctness` and `performance`.

The workload itself does nothing but randomly add `ReadConflictRange` and `WriteConflictRange` to a transaction and then `commit`. Consequently, when we have small keyspace and large `ConflictRange`, the transaction is more likely to be conflicting with others.

**Correctness**: 
For every `commit` caught by a `not_commited` error, we retrieve the reported `conflict_range` for the `commit`. Since the API is designed to return all the `read_conflict_ranges` conflicting with possibly committed `write_conflict_range` and union any overlapped ones, we assert that for each returned `KeyRange`, it contains at least one `readConflictRange` of which we randomly added for the transaction in this `commit`.

**Performance**: 
There are three versions of code testing based on `bdf31a7297`.
1.  With only the workload added
2.  Added the workload and the API, and turned off the option to report conflicts
3.  Added the workload and the API, and turned on the option to report conflicts

By intuition, the performance overhead between version `1` and `2` should be small enough to ignore.
And the overhead between `1` and `3` should be acceptable.

Different test configs are going to be tested, which includes:

- Size of key (Large&Regular size of keys)
- Conflicting rate (High&Regular conflicting rate)
- Number of Resolvers (Multiple&Single Resolvers)

zjuLcg commented:
A simple python script shows how to use the current API.

```
import threading
import fdb
import base64
import random
import json

fdb.api_version(620)
db = fdb.open("your_path_to_cluster_file")
// global var to control the key space
start=0
end=1000

def keyForIndex(ind):
   return "Prefix{0:*>10d}".format(ind).encode()

@fdb.transactional
def increment(tr):
    tr.options.set_report_conflicting_keys()
    if tr[b'\xff\xff/conflicting_keys/json'] != None:
        temp = tr[b'\xff\xff/conflicting_keys/json']
        jstr = json.loads(temp)
        print("\t".join(["start:{0};end:{1}".format(base64.b64decode(pair["begin"]), base64.b64decode(pair['end'])) for pair in jstr]))
    k1=random.randint(start, end)
    k2=random.randint(k1, end)
    key1=keyForIndex(k1)
    key2=keyForIndex(k2)
    tr.add_read_conflict_range(key1, key2)
    print("Read Conflict Range: ({0}, {1})".format(key1, key2))
    k3=random.randint(k1,k2)
    tr[keyForIndex(k3)] = b"Test"

@fdb.transactional
def clear(tr):
    key=keyForIndex(random.randint(start,end))
    # assert isinstance(key, bytes), "Key Must be bytes"
    del tr[key]

threads = []
for _ in range(1):
    for _ in range(10):
        threads.append(threading.Thread(target=clear, args=[db], daemon=True))
        threads[-1].start()
    threads.append(threading.Thread(target=increment, args=[db], daemon=True))
    threads[-1].start()

for thread in threads:
    thread.join()
```

zjuLcg commented:
Currently,  there are three feasible ways discussed to expose this API:
- JSON string through GET (Current Implementation)
To make things easier and avoid the compatibility issues with bindings. Currently, it is exposed as a simple `get` operation from a special key `\xff\xff/conflicting_keys/json`. And the conflicting keys are returned as a JSON string, formatted as an array of KeyRange.
In each KeyRange, it is formatted as `{"begin": ***, "end": ***}`.
The issue here is we may need to encode the keys to prevent the key bytes destroy the json format.
For example, `base64` can be used to encode the key values.
But the user need to decode it after calling the API. It may add some complexity here.

- Schema using flat-buffer through GET (The most general way to do)
Rather than using JSON, which needs to do the encoding. We can define a schema for the returned type of GET.

- KeyRange through getRange (Most clean way)
Same like using `GET`, here we call `getRange` using special keys like `getRange("\xff\xff/conflicting_keys", "\xff\xff/conflicting_keys\xff")` to get `RangeResult`.
The returned KV pairs can be formatted by two ways:
1. Key is the start key of the KeyRange and the value is the end of the KeyRange
2. Value is a boolean which represents whether keys greater or equal than the key is conflicting.

mpilman commented on fdbclient/MasterProxyInterface.h:
```diff
@@ -136,6 +137,10 @@ struct CommitTransactionRequest : TimedRequest {
 	}
+
+	void reportConflictingKeys(){
+		transaction.report_conflicting_keys = true;
```
I don't think the way you're setting this transaction option is consistent with other transaction options. I think the following would be more consistent (and therefore easier to understand for existing users):
1. Remove this method from `CommitTransactionRequest` - it is not needed.
1. Add `reportConflictingKeys` to the `TransactionOptions` in `NativeAPI`  and make sure is set back to `false` when reset is called on it.
1. In `Transaction::commitMutations()` you can check whether this option was set and then set the option in the transaction object directly.

zjuLcg replied:
Totally agree. My bad to forget this bad way to set it. Fixed it.

mpilman commented on fdbclient/ReadYourWrites.actor.cpp:
```diff
@@ -1279,6 +1279,14 @@ Future< Standalone<RangeResultRef> > ReadYourWritesTransaction::getRange(
 		}
 	}
 	
+	if (begin.getKey() == LiteralStringRef("\xff\xff/conflicting_keys") && end.getKey() == LiteralStringRef("\xff\xff/conflicting_keys\xff")) {
```
This is a bit weird: this will only allow me to fetch the whole list in one query. But I think the following should also work:

```
tr.getRange(LiteralStringRef("\xff\xff/conflicting_keys/bar"), LiteralStringRef("\xff\xff/conflicting_keys/bar"));
```

zjuLcg replied:
I have changed it to the way we discussed and add comments in the code.

mpilman commented on fdbclient/ReadYourWrites.actor.cpp:
```diff
@@ -1189,6 +1189,43 @@ ACTOR Future<Standalone<RangeResultRef>> getWorkerInterfaces (Reference<ClusterC
+		// Thus, we use a hack way here to achieve it.
+		// We create an empty RYWTransaction and write all conflicting key/values to it.
+		// Since it is RYWTr, we can call getRange on it with same parameters given to the original getRange
+		state ReadYourWritesTransaction hackTr(ryw->getDatabase());
```
I think for this to work you need to do two more things:

1. You need to set the read version explicitly - otherwise you will fetch a read version from the cluster.
1. You need to first clear the range (something like `hackTr.clear("\xff/conflicting_keys/", "\xff/conflicting_keys0")`).  Without that, the transaction has no way of knowing whether it knows about the whole range

zjuLcg replied:
Done. Thanks a lot.

mpilman commented on fdbclient/ReadYourWrites.actor.cpp:
```diff
@@ -1189,6 +1189,43 @@ ACTOR Future<Standalone<RangeResultRef>> getWorkerInterfaces (Reference<ClusterC
+		}
+		state KeySelector conflictingKeysBegin = KeySelectorRef(begin.getKey().withPrefix(conflictingKeysPrefix), begin.orEqual, begin.offset);
+		state KeySelector conflictingKeysEnd = KeySelectorRef(end.getKey().withPrefix(conflictingKeysPrefix), end.orEqual, end.offset);
+		Standalone<RangeResultRef> resultWithKeyPrefix = wait(hackTr.getRange(conflictingKeysBegin, conflictingKeysEnd, limits, snapshot, reverse));
```
```suggestion
		Future<Standalone<RangeResultRef>> resultWithKeyPrefixFuture = hackTr.getRange(conflictingKeysBegin, conflictingKeysEnd, limits, snapshot, reverse);
		ASSERT(resultWithKeyPrefixFuture.isReady());
		RangeResultRef resultWithKeyPrefix = resultWithKeyPrefixFuture.get();
```

zjuLcg replied:
Done. Thanks.

mpilman commented on fdbclient/ReadYourWrites.actor.cpp:
```diff
@@ -1189,6 +1189,43 @@ ACTOR Future<Standalone<RangeResultRef>> getWorkerInterfaces (Reference<ClusterC
 	}
 }
 
+ACTOR Future<Standalone<RangeResultRef>> getConflictingKeys(
```
This should not be an actor. Instead it should assert that it never needs to wait on anything (see below)

zjuLcg replied:
The code is already refactored as we discussed. Thanks.

zjuLcg replied:
The code is refactored as we discussed.

mpilman commented on fdbclient/MasterProxyInterface.h:
```diff
@@ -104,14 +104,15 @@ struct CommitID {
 	Version version; 			// returns invalidVersion if transaction conflicts
 	uint16_t txnBatchId;
 	Optional<Value> metadataVersion;
+	Optional<Standalone<VectorRef<KeyRangeRef>>> conflictingKeyRanges;
```
Sorry for this rather late suggestion: I think a lot of your code would become cleaner if this is a `ReadYourWritesTransaction`. It would also be faster as then you would only need to add all ranges to it once

mpilman commented on fdbserver/MasterProxyServer.actor.cpp:
```diff
@@ -1026,9 +1032,44 @@ ACTOR Future<Void> commitBatch(
-			trs[t].reply.sendError(not_committed());
+			// If enable the option to report conflicting keys from resolvers, we union all conflicting key ranges here and send back through CommitID
+			if (trs[t].transaction.report_conflicting_keys) {
+				Standalone<VectorRef<KeyRangeRef>> conflictingEntries;
```
This code is in a critical path - we need to be very careful to make this as CPU efficient as possible. I think we can optimize this a bit more, let's talk in person when you have some time.

zjuLcg replied:
Pushed this sorting and union process to client-side. Thanks.

mpilman commented on fdbserver/SkipList.cpp:
```diff
@@ -757,17 +760,33 @@ class SkipList : NonCopyable
 		Version version;
 		bool *result;
 		int state;
+		Standalone<VectorRef<KeyRangeRef>>* conflictingKeyRange; // null if report_conflicting_keys is not enabled.
```
Optional would probably be better here.
```suggestion
		Optional<Standalone<VectorRef<KeyRangeRef>>> conflictingKeyRange; // null if report_conflicting_keys is not enabled.
```

zjuLcg commented on fdbclient/NativeAPI.actor.cpp:
```diff
@@ -2673,6 +2674,29 @@ ACTOR static Future<Void> tryCommit( Database cx, Reference<TransactionLogInfo>
+						tr->info.conflictingKeysRYW = std::make_shared<ReadYourWritesTransaction>(tr->getDatabase());
+						// To make the getRange call local, we need to explicitly set the read version here.
+						// This version number 100 set here does nothing but prevent getting read version from the proxy
+						tr->info.conflictingKeysRYW.get()->setVersion(100);
```
Thanks. Did not know that. Done.

zjuLcg commented on fdbclient/NativeAPI.actor.h:
```diff
@@ -132,19 +132,24 @@ struct TransactionOptions {
+	std::shared_ptr<ReadYourWritesTransaction> conflictingKeysRYW;
 
-	explicit TransactionInfo( TaskPriority taskID ) : taskID(taskID), useProvisionalProxies(false) {}
+	explicit TransactionInfo( TaskPriority taskID ) : taskID(taskID), useProvisionalProxies(false), conflictingKeysRYW(nullptr) {}
```
Done. Thanks.

zjuLcg commented on fdbclient/ReadYourWrites.actor.cpp:
```diff
@@ -1279,6 +1279,36 @@ Future< Standalone<RangeResultRef> > ReadYourWritesTransaction::getRange(
+		begin.setKey(beginConflictingKey);
+		end.setKey(endConflictingKey);
+		if (tr.info.conflictingKeysRYW) {
+			Future<Standalone<RangeResultRef>> resultWithKeyPrefixFuture = tr.info.conflictingKeysRYW.get()->getRange(begin, end, limits, snapshot, reverse);
```
Done. Thanks.

mpilman commented on fdbclient/NativeAPI.actor.h:
```diff
@@ -132,19 +132,24 @@ struct TransactionOptions {
 	bool useProvisionalProxies;
+	// Used to save conflicting keys if FDBTransactionOptions::REPORT_CONFLICTING_KEYS is enabled
+	// shared_ptr used here since TransactionInfo is sometimes copied as function parameters.
+	std::shared_ptr<ReadYourWritesTransaction> conflictingKeysRYW;
```
```suggestion
	Reference<ReadYourWritesTransaction> conflictingKeysRYW;
```
Reference should be faster than shared_ptr

zjuLcg replied:
As we discussed, `Reference` is better to be used here. 
But one problem is that `Reference` needs the underlying type to have `delref` function. 
However, since we use forward declaration here to avoid cyclic dependency.
It does not compile since the compiler knows nothing about `ReadYourWritesTransaction` here.

zjuLcg replied:
Discussed in person. Keep it as `share_ptr` for now. May solve the issue in the future.

ajbeamon replied:
Could this be solved by forward declaring `TransactionInfo` here and defining it the cpp file? I'm not sure it's used anywhere else.

mpilman replied:
I think that would be one option. But we decided to solve the whole problem differently which will get rid of this member altogether.

mpilman commented on fdbserver/ResolverInterface.h:
```diff
@@ -77,10 +77,11 @@ struct ResolveTransactionBatchReply {
 	VectorRef<uint8_t> committed;
 	Optional<UID> debugID;
 	VectorRef<VectorRef<StateTransactionRef>> stateMutations;  // [version][transaction#] -> (committed, [mutation#])
+	std::map<int, Standalone<VectorRef<KeyRangeRef>>> conflictingKeyRangeMap; // transaction index -> conflicting keyRanges given by the resolver
```
```suggestion
	std::map<int, VectorRef<KeyRangeRef>> conflictingKeyRangeMap; // transaction index -> conflicting keyRanges given by the resolver
```
I think this doesn't have to be standalone - you can use the Arena instead

zjuLcg replied:
As we discussed before, I changed the code to use the index instead of keys.
I changed the data structure to `std::vector` here, which is easier to use.
Also, the `arena` in `ResolveTransactionBatchReply` is changed somewhere in the proxy code.
Since I have not found the place where the change happens, I just avoid using the `arena` here.

mpilman commented on fdbserver/workloads/ReportConflictingKeys.actor.cpp:
```diff
@@ -0,0 +1,213 @@
+				if (!self->skipCorrectnessCheck && self->reportConflictingKeys && isConflict) {
+					state KeyRange ckr = KeyRangeRef(LiteralStringRef("\xff\xff/conflicting_keys/"), LiteralStringRef("\xff\xff/conflicting_keys/\xff"));
+					// The getRange here using the special key prefix "\xff\xff/conflicting_keys/" happens locally, so error handling is not needed here
+					Standalone<RangeResultRef> conflictingKeyRanges = wait(tr.getRange(ckr, readConflictRanges.size() * 2));
```
```suggestion
                                       Future<Standalone<RangeResultRef>> conflictingKeyRangesF = tr.getRange(ckr, readConflictRanges.size() * 2);
					ASSERT(conflictingKeyRangesF.isReady());
					Standalone<RangeResultRef>> conflictingKeyRanges = conflictingKeyRangesF.get();
```

zjuLcg replied:
Done. Thanks.

mpilman commented on fdbclient/ReadYourWrites.actor.cpp:
```diff
@@ -1279,6 +1279,36 @@ Future< Standalone<RangeResultRef> > ReadYourWritesTransaction::getRange(
+	// <key2> : '0' - any keys equal or larger than this key are (definitely) not conflicting keys
+	// Due to the implementation of resolver, currently, 
+	// we can only give key ranges that contain at least one key is conflicting.
+	Key conflictingKeysPreifx = LiteralStringRef("\xff\xff/conflicting_keys/");
```
```suggestion
	Key conflictingKeysPreifx = LiteralStringRef("\xff\xff/transaction/conflicting_keys/");
```

zjuLcg replied:
Done. Thanks.

zjuLcg commented on fdbclient/NativeAPI.actor.cpp:
```diff
@@ -2673,6 +2674,29 @@ ACTOR static Future<Void> tryCommit( Database cx, Reference<TransactionLogInfo>
+						// Consequently, we use a hack way here to achieve it.
+						// We create an empty RYWTransaction and write all conflicting key/values to it.
+						// Since it is RYWTr, we can call getRange on it with same parameters given to the original getRange.
+						tr->info.conflictingKeysRYW = std::make_shared<ReadYourWritesTransaction>(tr->getDatabase());
```
DIscussed in person. Solved.

mpilman commented:
test his please

mpilman commented:
@fdb-build test cmake linux please

mpilman commented on fdbclient/ReadYourWrites.actor.cpp:
```diff
@@ -1279,6 +1279,38 @@ Future< Standalone<RangeResultRef> > ReadYourWritesTransaction::getRange(
+				tr.info.conflictingKeysRYW->getRange(begin, end, limits, snapshot, reverse);
+			// Make sure it happens locally
+			ASSERT(resultWithKeyPrefixFuture.isReady());
+			return resultWithKeyPrefixFuture.get();
```
You have to add the prefix back to the keys here.

zjuLcg replied:
Sorry for the bad naming. I forgot to change it.
The results are with no prefix right now.
The current behavior is when you call `getRagne("\xff\xff/transaction/conflicting_keys/<start_key>, ", "\xff\xff/transaction/conflicting_keys/<end_key>, ",)`
The keys returned are not with prefix like:
`far: 1; fool: 0`
I think it is enough for now since currently there is only conflicting keys right now.

zjuLcg replied:
As we discussed, I have added the prefix to the returned keys, which will keep the consistency of `getRange`.

mpilman commented on fdbserver/ResolverInterface.h:
```diff
@@ -77,10 +77,11 @@ struct ResolveTransactionBatchReply {
 	template <class Archive>
 	void serialize(Archive& ar) {
-		serializer(ar, committed, stateMutations, arena, debugID);
+		serializer(ar, committed, stateMutations, arena, debugID, conflictingKeyRangeMap);
```
Can you try to make `conflictingKeyRangeMap` a map of `VectorRef` again and then move the arena to the end (ie make it the last argument to `serializer`)? This might fix the problem you saw before.

Arenas and serializers are a bit weird, and with flatbuffers Arena is the only type you are allowed to move around without breaking wire-compatability.

zjuLcg replied:
It solved the problem!
Changed back to VectorRef.

etschannen commented on fdbclient/NativeAPI.actor.cpp:
```diff
@@ -2742,6 +2743,33 @@ ACTOR static Future<Void> tryCommit( Database cx, Reference<TransactionLogInfo>
+						// merge duplicate indices
+						const auto cKRs = ci.conflictingKRIndices.get();
+						std::set<int> mergedIds(cKRs.begin(), cKRs.end());
+						for (auto const & rCRIndex : mergedIds) {
```
This implementation is relying on the read conflict ranges not overlapping. This is true when the read conflicts come from ReadYourWrites transactions, but not necessarily true if someone is using NativeAPI directly or has turned on READ_YOUR_WRITES_DISABLE. You should keep track of the largest end key you have seen, and if the beginning of a range is less than that, do a much more complicated merge (I think you should be able to use `krmSetRange`).

etschannen replied:
you should modify your test so it can detect this bug

zjuLcg replied:
Done.
The transactions in the test workload are randomly enabled `READ_YOUR_WRITES_DISABLE ` option.
Since `READ_YOUR_WRITES_DISABLE` does the same as native `transaction`'s `addReadConflictRange` and `addWriteConflictRange`, we only need to test one of them here.

etschannen commented on fdbserver/workloads/ReportConflictingKeys.actor.cpp:
```diff
@@ -0,0 +1,222 @@
+						Key endKey = endKeyWithPrefix.key.removePrefix(conflictingKeysPreifx);
+						KeyRangeRef kr = KeyRangeRef(startKey, endKey);
+						if (!std::any_of(readConflictRanges.begin(), readConflictRanges.end(), [&kr](KeyRange rCR) {
+							    // Read_conflict_range remains same in the resolver.
```
This check is only determining if the reported read conflict range was one of the ones we added to the transaction, not whether the range was actually the cause of the conflict. I would suggest a more controlled test where you start two transactions with the same read version, commit one of them first, then commit the other. If the second commit does not fail, you can verify that the read conflict ranges of the second transactions do not overlap with the write conflict ranges of the first transaction. If the second commit does fail than you can check that all reported conflict ranges overlap with write conflict ranges from the first transaction.

etschannen replied:
Look at the ConflictRange test, because it does something similar. You will also want to do the same thing that test does, and in the Test.txt file disable buggify and connection failures (connectionFailuresDisableDuration=100000
buggify=off)

zjuLcg replied:
Done. It now does what you describe here.

>start two transactions with the same read version, commit one of them first, then commit the other. If the second commit does not fail, you can verify that the read conflict ranges of the second transactions do not overlap with the write conflict ranges of the first transaction. If the second commit does fail than you can check that all reported conflict ranges overlap with write conflict ranges from the first transaction.

zjuLcg commented on fdbclient/KeyRangeMap.actor.cpp:
```diff
@@ -216,3 +216,70 @@ ACTOR Future<Void> krmSetRangeCoalescing( Transaction *tr, Key mapPrefix, KeyRan
 	return Void();
 }
+
+ACTOR Future<Void> krmSetRangeCoalescing( Reference<ReadYourWritesTransaction> tr, Key mapPrefix, KeyRange range, KeyRange maxRange, Value value ) {
```
Done. Thanks.

zjuLcg commented on fdbclient/NativeAPI.actor.cpp:
```diff
@@ -2742,6 +2743,36 @@ ACTOR static Future<Void> tryCommit( Database cx, Reference<TransactionLogInfo>
+						tr->info.conflictingKeysRYW->set(conflictingKeysPrefix, conflictingKeysFalse);
+						// drop duplicate indices and merge overlapped ranges
+						// Note: addReadConflictRange in native transaction object does not merge overlapped ranges
+						state std::set<int> mergedIds(conflictingKRIndices.begin(), conflictingKRIndices.end());
```
Done. Thanks.

zjuLcg commented on fdbserver/workloads/ReportConflictingKeys.actor.cpp:
```diff
@@ -0,0 +1,237 @@
+					    tr2.getRange(ckr, readConflictRanges.size() * 2);
+					ASSERT(conflictingKeyRangesFuture.isReady());
+					const Standalone<RangeResultRef> conflictingKeyRanges = conflictingKeyRangesFuture.get();
+					ASSERT(conflictingKeyRanges.size() && (conflictingKeyRanges.size() % 2 == 0));
```
Done. Thanks.

zjuLcg commented on fdbclient/NativeAPI.actor.cpp:
```diff
@@ -2742,6 +2743,36 @@ ACTOR static Future<Void> tryCommit( Database cx, Reference<TransactionLogInfo>
+							const KeyRange kr = req.transaction.read_conflict_ranges[rCRIndex];
+							wait(krmSetRangeCoalescing(hackTr, conflictingKeysPrefix, kr, allKeys, conflictingKeysTrue));
+						}
+						hackTr.extractPtr(); // Avoid the Reference to destroy the RYW object
```
Done. Add a try loop here to deal with the case.

atn34 commented on fdbclient/KeyRangeMap.h:
```diff
@@ -103,6 +103,7 @@ void krmSetPreviouslyEmptyRange( struct CommitTransactionRef& tr, Arena& trArena
 Future<Void> krmSetRange( Transaction* const& tr, Key const& mapPrefix, KeyRange const& range, Value const& value );
 Future<Void> krmSetRange( Reference<ReadYourWritesTransaction> const& tr, Key const& mapPrefix, KeyRange const& range, Value const& value );
 Future<Void> krmSetRangeCoalescing( Transaction* const& tr, Key const& mapPrefix, KeyRange const& range, KeyRange const& maxRange, Value const& value );
+Future<Void> krmSetRangeCoalescing( Reference<ReadYourWritesTransaction> const& tr, Key const& mapPrefix, KeyRange const& range, KeyRange const& maxRange, Value const& value );
```
I forgot that we added this new overload of krmSetRangeCoalescing. I think it should just take a pointer like the other one, and then we don't need to do the whole "holdWhile" thing

zjuLcg replied:
I made it as a Reference like what `krmSetRange` does to make the interface consistent.

atn34 replied:
Ahh right. Fair enough.

xumengpanda commented on tests/fast/ReportConflictingKeys.txt:
```diff
@@ -0,0 +1,10 @@
+    readConflictRangeCountPerTx=10
+    writeConflictRangeCountPerTx=10
+    connectionFailuresDisableDuration=100000
+    buggify=off
```
Why do we have to turn off buggify for this?

mpilman replied:
that's an excellent question. @zjuLcg can we remove this line?

atn34 replied:
I think it's intentional, see https://github.com/apple/foundationdb/pull/2257#discussion_r390035115. My guess is that it's not strictly necessary, but it improves coverage since we tr.commit() failing with say commit_unknown_result is not interesting here.

zjuLcg replied:
I think it follows the same way as `ConflictRange` workload.

mpilman replied:
What Andrew is saying makes sense - though even with buggify on I would expect most transactions would not fail with `commit_unknown_result`...

I wonder whether this causes some code-paths not getting tested well enough. I see that @etschannen added this to `ConflictRangeCheck.txt` ~2 years ago. Evan, can you share your opinion on this.

sfc-gh-gkliot commented:
Was the transaction option report_conflicting_keys added to Java Bindings API?

sfc-gh-gkliot commented:
Docs in https://github.com/apple/foundationdb/blob/release-6.3/documentation/sphinx/source/special-keys.rst#transaction-module  talk about \xff\xff/transaction/read_conflict_range/, not /transaction/conflicting_keys/
This is confusing.

sfc-gh-anoyes commented:
> Docs in https://github.com/apple/foundationdb/blob/release-6.3/documentation/sphinx/source/special-keys.rst#transaction-module talk about \xff\xff/transaction/read_conflict_range/, not /transaction/conflicting_keys/
> This is confusing.

The linked docs describe `\xff\xff/transaction/read_conflict_range/`, `\xff\xff/transaction/write_conflict_range/`, and `\xff\xff/transaction/conflicting_keys/`. These are 3 different concepts related to conflict ranges that you can inspect for your transactions

sfc-gh-gkliot commented:
Ohh, right. I see. Thank you!
